### PR TITLE
Feature: cross validate timings

### PR DESCRIPTION
**Pull request by alekseykalyagin**
## Description

Closes issue https://github.com/MobileTeleSystems/RecTools/issues/138

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Optimization

## How Has This Been Tested?
Before submitting a PR, please check yourself against the following list. It would save us quite a lot of time.
- Have you read the contribution guide?
- Have you updated the relevant docstrings? We're using Numpy format, please double-check yourself
- Does your change require any new tests?
- Have you updated the changelog file?

### CODE REVIEW

File: rectools/model_selection/cross_validate.py [hunk 1]
```diff
@@ -100,6 +124,47 @@ def cross_validate(  # pylint: disable=too-many-locals
             ]
         }
     """
+
+    def fit_recommend(model: ModelBase, ref_model: bool = False) -> tp.Tuple[pd.DataFrame, tp.Dict[str, tp.Any]]:
```
[the remark this answers is not in the document]
Comment (alekseykalyagin) [reply rectools/model_selection/cross_validate.py]: done

File: rectools/model_selection/cross_validate.py [hunk 2]
```diff
@@ -123,28 +145,16 @@ def cross_validate(  # pylint: disable=too-many-locals
...
-            model.fit(fold_dataset)
-            ref_reco[model_name] = model.recommend(
-                users=test_users,
-                dataset=fold_dataset,
-                k=k,
-                filter_viewed=filter_viewed,
-                items_to_recommend=items_to_recommend,
-                on_unsupported_targets=on_unsupported_targets,
-            )
+            ref_timings: tp.Optional[tp.Dict[str, float]] = {} if compute_timings and validate_ref_models else None
```
[the remark this answers is not in the document]
Comment (alekseykalyagin) [reply rectools/model_selection/cross_validate.py]: done

File: rectools/model_selection/cross_validate.py [hunk 3]
```diff
@@ -123,28 +145,16 @@ def cross_validate(  # pylint: disable=too-many-locals
 
         # ### Train ref models if any
         ref_reco = {}
+        ref_res = {}
```
[the remark this answers is not in the document]
Comment (alekseykalyagin) [reply rectools/model_selection/cross_validate.py]: done

File: rectools/model_selection/cross_validate.py [hunk 4]
```diff
@@ -153,6 +163,33 @@ def cross_validate(  # pylint: disable=too-many-locals
...
+                    reco = model.recommend(
+                        users=test_users,
+                        dataset=fold_dataset,
+                        k=k,
+                        filter_viewed=filter_viewed,
+                        items_to_recommend=items_to_recommend,
+                        on_unsupported_targets=on_unsupported_targets,
+                    )
+
+                model_res = timings or {}
```
[the remark this answers is not in the document]
Comment (alekseykalyagin) [reply rectools/model_selection/cross_validate.py]: done

File: tests/model_selection/test_cross_validate.py [hunk 5]
```diff
@@ -371,5 +371,367 @@ def test_happy_path_with_intersection(
...
+            ref_models=ref_models,
+            validate_ref_models=validate_ref_models,
+            compute_timings=compute_timings,
+        )
+
+        expected_keys = {"fit_time", "recommend_time"}
+
+        if compute_timings:
+            for data in actual["metrics"]:
+                assert len(expected_keys.intersection(set(data.keys()))) == 2
```
[the remark this answers is not in the document]
Comment (alekseykalyagin) [reply tests/model_selection/test_cross_validate.py]: done, chose threshold = 0.5

File: tests/model_selection/test_cross_validate.py [hunk 6]
```diff
@@ -371,5 +371,172 @@ def test_happy_path_with_intersection(
             ],
             "metrics": expected_metrics,
         }
+        assert actual == expected
 
+    @pytest.mark.parametrize(
+        "validate_ref_models,expected_metrics,compute_timings",
```
Comment (blondered) [inline tests/model_selection/test_cross_validate.py:377]: Let's improve it more and make it more clear.
You have `validate_ref_models` and `expected_metrics` depending on each other.
And you have `compute_timings` that has nothing to do with them. But you duplicate same values of previous parameters.
To stop duplicating code just move  `compute_timings` to different `@pytest.mark.parametrize` decorator.

Comment (alekseykalyagin) [reply tests/model_selection/test_cross_validate.py]: done

File: tests/model_selection/test_cross_validate.py [hunk 7]
```diff
@@ -371,5 +371,247 @@ def test_happy_path_with_intersection(
             ],
             "metrics": expected_metrics,
         }
+        assert actual == expected
 
+    @pytest.mark.parametrize(
+        "ref_models,validate_ref_models,expected_metrics,compute_timings",
+        (
+            (
+                ["popular"],
```
[the remark this answers is not in the document]
Comment (alekseykalyagin) [reply tests/model_selection/test_cross_validate.py]: done

File: tests/model_selection/test_cross_validate.py [hunk 8]
```diff
@@ -371,5 +371,247 @@ def test_happy_path_with_intersection(
...
+                        "fit_time": 0.0,
+                        "recommend_time": 0.0,
+                    },
+                    {
+                        "model": "random",
+                        "i_split": 1,
+                        "precision@2": 0.375,
+                        "recall@1": 0.5,
+                        "intersection_popular": 0.75,
+                        "fit_time": 0.0,
```
[the remark this answers is not in the document]
Comment (alekseykalyagin) [reply tests/model_selection/test_cross_validate.py]: done

File: rectools/model_selection/cross_validate.py [hunk 9]
```diff
@@ -24,6 +25,26 @@
...
+    ----------
+    label : str
+        Label to store the timing result in the timings dictionary.
+    timings : dict, optional
+        Dictionary to store the timing results. If None, timing is not recorded.
+    """
+    if timings is not None:
+        start_time = time.time()
+        yield
+        timings[label] = round(time.time() - start_time, 2)
```
[the remark this answers is not in the document]
Comment (alekseykalyagin) [reply rectools/model_selection/cross_validate.py]: done

File: rectools/model_selection/cross_validate.py [hunk 10]
```diff
@@ -36,6 +57,7 @@ def cross_validate(  # pylint: disable=too-many-locals
     ref_models: tp.Optional[tp.List[str]] = None,
     validate_ref_models: bool = False,
     on_unsupported_targets: ErrorBehaviour = "warn",
+    compute_timings: bool = False,
```
Comment (blondered) [inline rectools/model_selection/cross_validate.py:60]: Please add new argument to docstring

File: rectools/model_selection/cross_validate.py [hunk 11]
```diff
@@ -24,6 +25,26 @@
...
+    """
+    if timings is not None:
+        start_time = time.time()
+        yield
+        timings[label] = round(time.time() - start_time, 5)
+    else:
+        yield
+
+
 def cross_validate(  # pylint: disable=too-many-locals
```
Comment (blondered) [inline rectools/model_selection/cross_validate.py:48]: Please update CHANGELOG.MD

File: rectools/model_selection/cross_validate.py [hunk 12]
```diff
@@ -24,6 +25,26 @@
 from .splitter import Splitter
 
 
+@contextmanager
+def compute_timing(label: str, timings: tp.Optional[tp.Dict[str, float]] = None) -> tp.Iterator[None]:
```
Comment (feldlime) [inline rectools/model_selection/cross_validate.py:29]: It seems this function accepts the `label` `timings` param but don't really need it

Let's please rewrite it in one of the following ways:
1. Remove both params and simply return the elapsed time without dictionaries
2. Rewrite it as a class

I personally prefer the second option since it's clearer.  
But anyway let's not use this labels and dictionary inside. We can easily fill them out of the class

And example (it's simplified a bit, please add init if required by linters, also types)
```
class Timer:        
    def __enter__(self):
        self._start = time.perf_counter()
        self._end = None
        return self

    def __exit__(self, *args):
        self._end = time.perf_counter()

    @property
    def elapsed(self):
        return self._end - self._start
    
    
with Timer() as timer:
    # code
    pass

fit_time = timer.elapsed
```

File: rectools/model_selection/cross_validate.py (hunk 10 above)
Comment (feldlime) [reply rectools/model_selection/cross_validate.py:60]: do we really need this param? what's wrong if we always measure the time?

File: rectools/model_selection/cross_validate.py [hunk 13]
```diff
@@ -24,6 +25,26 @@
...
+
+    Parameters
+    ----------
+    label : str
+        Label to store the timing result in the timings dictionary.
+    timings : dict, optional
+        Dictionary to store the timing results. If None, timing is not recorded.
+    """
+    if timings is not None:
+        start_time = time.time()
```
Comment (feldlime) [inline rectools/model_selection/cross_validate.py:41]: Please use `time.perf_counter` instead, it's more correct for measuring time intervals

File: rectools/model_selection/cross_validate.py [hunk 14]
```diff
@@ -24,6 +25,26 @@
...
+        yield
+        timings[label] = round(time.time() - start_time, 5)
```
Comment (feldlime) [inline rectools/model_selection/cross_validate.py:43]: Please don't do this

1. If we need to format the values somehow it should always be done separately. We should separate the computing level and presentation level. From the computing level we should always return the raw values.
2. In this specific case I think we shouldn't format the value at all. I don't see much sense in it, also we're not doing this for other metrics.

**Review by feldlime — CHANGES_REQUESTED** [review]
